File: cpp/lv0/옹알이/옹알이.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// 1. string의 모든 원소를 돌면서 첫 단어에 "aya", "ye", "woo", "ma"가 포함되어 있는지 확인
// 2. 첫 단어에 포함되어 있다면, 이어지는 단어에 위 4 단어가 포함되어 있는지 확인
// 3. 위 두 가지 경우를 만족한다면 answer++



int solution(vector<string> babbling) {
    std::vector<string> v{"aya", "ye", "woo", "ma"};
    
    int answer = 0;
    
    for (int i = 0; i < babbling.size(); ++i) {
        int j = 0;
        std::string temp = babbling[i];
        while (true) {
            // 첫 문자열이 v 원소와 같은지 파악 -> 같다면
            if (temp.rfind(v[j], 0) == 0) {

                // 같은 문자열을 제외하고 뒤부터 temp에 담음
                temp = temp.substr(v[j].length());

                // temp에 담은 결과값이 0이면 answer값을 올리고 break
                if (temp.length() == 0) {
                    ++answer;
                    break;
                }
                j = 0;
                continue;
            }
            else {
                ++j;
                if (j > v.size() - 1) {
                    break;
                }
            }
        }
    }
    return answer;
}


int main() {
    std::vector<string> v{"aya", "yee", "u", "maa", "wyeoo"};
    std::cout << solution(v) << std::endl;

    // std::string s {"hihihihello"};

    // std::string s1 = "hihihihello";

    // std::string s0 = s.substr(s1.length());

    // std::cout << s0.length() << std::endl;
}

File: go/lv0/옹알이/옹알이.go
import "strings"

func solution(babbling []string) int {
    result := 0
    // 발음할 수 있는 단어를 담은 slice 생성
    words := []string {"aya", "ye", "woo", "ma"}
    
    for _, value := range babbling {
        i := 0
        // 현재 보고있는 value를 temp라는 변수에 할당
        temp := value
        // 문자열의 접두어가 words중 포함되어있는지 확인
        for {
            // 만약 있다면
            if strings.HasPrefix(temp, words[i]) {
                
                // 접두어의 길이만큼 temp를 앞에서부터 잘라냄
                temp = temp[len(words[i]):]
                // 잘라냈을 때, 길이가 0이라면
                if len(temp) == 0 {
                    // 발음할 수 있는 것이기 때문에 result값을 ++함
                    result++
                    // 반복문 종료
                    break
                }
                // 길이가 0이 아니라면, 남은 잘려진 temp 문자열에 대해
                // word[0]번 요소부터 다시 검사 시작
                i = 0
                
            // 포함되어 있지 않다면
            } else {
                // i를 증가해가며 word[i]가 temp의 접두어인지 확인
                i++
                // 만약 i가 words slice의 길이를 초과한다면 break
                if i > len(words) - 1 {
                    break
                }
            }
        }
    }
    return result
}